tests/test_logic.py

```diff
@@ -531,6 +532,41 @@ def test_Controller_on_sync_success(homedir, config, mocker):
     co.file_missing.emit.assert_called_once_with(missing.source.uuid, missing.uuid, str(missing))
 
 
+def test_Controller_on_sync_success_when_current_user_deleted(mocker, homedir):
+    co = Controller("http://localhost", mocker.MagicMock(), mocker.MagicMock(), homedir)
+
+    class DeletedUser(db.User):
+        @property
+        def username(self):
+            raise sqlalchemy.orm.exc.ObjectDeletedError(attributes.instance_state(db.User()), None)
+
+    co.authenticated_user = DeletedUser()
+    co.api = "not None"
+    co.gui = mocker.MagicMock()
+
+    co.on_sync_success()
+
+    assert co.authenticated_user is None
+    assert co.api is None
+    assert co.is_authenticated is False
+    co.gui.logout.assert_called_once_with()
+
+
+def test_Controller__close_client_session(mocker, homedir):
+    co = Controller("http://localhost", mocker.MagicMock(), mocker.MagicMock(), homedir)
+    co.authenticated_user = factory.User(username="foo")
+    co.api = "not None"
+    co.gui = mocker.MagicMock()
+
+    co._close_client_session()
+
+    assert co.authenticated_user is None
+    assert co.api is None
+    assert co.is_authenticated is False
+    co.gui.logout.assert_called_once_with()
+    co.gui.show_login.assert_called_once_with(error="Your session expired. Please log in again.")
```

Comment:
Should we also test that the token was invalidated?

```suggestion
    co.gui.invalidate_token.assert_called_once_with()
    co.gui.logout.assert_called_once_with()
    co.gui.show_login.assert_called_once_with(error="Your session expired. Please log in again.")
```

Reply:
When we close the client session, we just set `co.api` and `co.authenticated_user` to None, so that the user can no longer make requests from the client. The log out forces the user to re-authenticate, so the client retrieves a new token. We could change this in the future to instead set `co.api.token` to None, but that goes against the existing implementation in the client for invalidating a session, and changes this from simply fixing the bug to fixing the bug _and_ changing the way we force a session restart.

Reply:
I'm asking because I saw it in:
https://github.com/freedomofpress/securedrop-client/pull/1412/files#diff-bd9c8ec560911b1924ac2bee334a948c32c82130491940b48a5dd2c38ec8110fR678

Reply:
Makes sense why you were asking. That code ultimately just sets the api and the auth'd user to None, which the test asserts on already, so I think we're good?

Reply:
Yes, as I said below, I don't think the extra assertion would be bringing much, and what you're saying makes sense :+1: